validator/src/journal/publisher.rs

```diff
@@ -1,4 +1,107 @@
-use std::collections::VecDeque;
+use proto::batch::Batch;
+use std::collections::{HashSet, VecDeque};
+use std::slice::Iter;
+
+/// Ordered batches wating to be processed
+pub struct PendingBatchesPool {
+    batches: Vec<Batch>,
+    ids: HashSet<String>,
+    limit: QueueLimit,
+}
+
+impl PendingBatchesPool {
+    pub fn new(sample_size: usize, initial_value: i32) -> PendingBatchesPool {
+        PendingBatchesPool {
+            batches: Vec::new(),
+            ids: HashSet::new(),
+            limit: QueueLimit::new(sample_size, initial_value),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.batches.len()
+    }
+
+    pub fn iter(&self) -> Iter<Batch> {
+        self.batches.iter()
+    }
+
+    fn contains(&self, id: &str) -> bool {
+        self.ids.contains(id)
+    }
+
+    fn reset(&mut self) {
+        self.batches = Vec::new();
+        self.ids = HashSet::new();
+    }
+
+    pub fn append(&mut self, batch: Batch) {
+        if self.contains(&batch.get_header_signature()) {
+            self.ids.insert(batch.get_header_signature().into());
+            self.batches.push(batch);
+        }
+    }
+
+    /// Recomputes the list of pending batches
+    ///
+    /// Args:
+    ///   committed (List<Batches>): Batches committed in the current chain
+    ///   since the root of the fork switching from.
+    ///   uncommitted (List<Batches): Batches that were committed in the old
+    ///   fork since the common root.
+    pub fn rebuild(&mut self, committed: Option<Vec<Batch>>, uncommitted: Option<Vec<Batch>>) {
+        let committed_set = committed
+            .unwrap_or(vec![])
+            .iter()
+            .map(|i| i.get_header_signature().into())
+            .collect::<HashSet<String>>();
+
+        let previous_batches = self.batches.clone();
+
+        self.reset();
+
+        // Uncommitted and pending are disjoint sets since batches can only be
+        // committed to a chain once.
+        for batch in uncommitted.unwrap_or(vec![]) {
+            if !committed_set.contains(batch.get_header_signature()) {
+                self.append(batch);
+            }
+        }
+
+        for batch in previous_batches {
+            if !committed_set.contains(batch.get_header_signature()) {
+                self.append(batch);
+            }
+        }
+    }
+
+    pub fn update(&mut self, mut still_pending: Vec<Batch>, last_sent: Batch) {
+        let last_index = self.batches
+            .iter()
+            .position(|i| i.get_header_signature() == last_sent.get_header_signature());
+
+        let unsent = if let Some(idx) = last_index {
+            still_pending.extend_from_slice(self.batches.clone().split_off(idx + 1).as_slice());
+            still_pending
+        } else {
+            self.batches.clone()
```

Comment:
This is going to be a very expensive clone, as this list can have as many as 100K batches in it at once. Instead, you should use `std::mem::swap` (https://doc.rust-lang.org/std/mem/fn.swap.html) to take the batch list out and replace it with an empty one.